agent/engine/task_manager.go

```diff
@@ -1045,16 +1049,81 @@ func (mtask *managedTask) containerNextState(container *apicontainer.Container)
+//
+// Next status is determined by whether the known and desired statuses are
+// equal, the next numerically greater (iota-wise) status, and whether
+// dependencies are fully resolved.
 func (mtask *managedTask) resourceNextState(resource taskresource.TaskResource) *resourceTransition {
+	resKnownStatus := resource.GetKnownStatus()
+	resDesiredStatus := resource.GetDesiredStatus()
+
+	if resKnownStatus == resDesiredStatus {
+		seelog.Debugf("Managed task [%s]: task resource [%s] at desired status: %s",
+			mtask.Arn, resource.GetName(), resource.StatusString(resDesiredStatus))
+
+		// Set nextState to ResourceStatusNone, so the knownState will not be updated
+		return &resourceTransition{
+			nextState:      resourcestatus.ResourceStatusNone,
+			status:         resource.StatusString(resourcestatus.ResourceStatusNone),
+			actionRequired: false,
+			reason:         dependencygraph.ResourcePastDesiredStatusErr,
+		}
+	}
+
+	if resKnownStatus > resDesiredStatus {
+		seelog.Debugf("Managed task [%s]: task resource [%s] has already transitioned beyond desired status(%s): %s",
+			mtask.Arn, resource.GetName(), resource.StatusString(resDesiredStatus), resource.StatusString(resKnownStatus))
+		return &resourceTransition{
+			nextState:      resourcestatus.ResourceStatusNone,
+			status:         resource.StatusString(resourcestatus.ResourceStatusNone),
+			actionRequired: false,
+			reason:         dependencygraph.ResourcePastDesiredStatusErr,
+		}
+	}
+	if err := dependencygraph.TaskResourceDependenciesAreResolved(resource, mtask.Containers); err != nil {
+		seelog.Debugf("Managed task [%s]: can't apply state to resource [%s] yet due to unresolved dependencies: %v",
+			mtask.Arn, resource.GetName(), err)
+		return &resourceTransition{
+			nextState:      resourcestatus.ResourceStatusNone,
+			status:         resource.StatusString(resourcestatus.ResourceStatusNone),
+			actionRequired: false,
+			reason:         err,
+		}
+	}
+
+	var nextState resourcestatus.ResourceStatus
 	if resource.DesiredTerminal() {
 		nextState := resource.TerminalStatus()
-		return &resourceTransition{
-			nextState:      nextState,
-			status:         resource.StatusString(nextState),
-			actionRequired: false, // since resource cleanup will be done while sweeping task
+		if !resource.KnownCreated() {
+			// If the resource's creation is in progress, we will not transit resource state before
+			// it finishes. Otherwise, the resource may not be cleaned up.
+			if resource.GetAppliedStatus() == resourcestatus.ResourceCreated {
+				nextState = resourcestatus.ResourceStatusNone
+			}
+			seelog.Debugf("set next state to %s", resource.StatusString(nextState))
+			return &resourceTransition{
+				nextState:      nextState,
+				status:         resource.StatusString(nextState),
+				actionRequired: false,
+			}
+		}
+		// If the resource does not have dependency on task network, it will be cleaned up while sweeping task
+		if !resource.DependOnTaskNetwork() {
```

Comment:
irrespective of task network dependency, today the resources are cleaned up only during sweep task. why is this check/logic needed?